subprojects/core/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRolesForMigration.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+/**
+ * The roles here all meant to be temporary roles used for migration only, to be removed in Gradle 9.0.
+ * <p>
+ * While we currently (Gradle 8.x) have to support the legacy behavior of configurations, we want to encode
+ * the knowledge of what the minimally required intended behavior is, so that we easily migrate to the
+ * {@link #eventualRole} in Gradle 9.0.
+ */
+public enum ConfigurationRolesForMigration implements ConfigurationRole {
+    @Deprecated
+    LEGACY_TO_INTENDED_RESOLVABLE_BUCKET(ConfigurationRoles.LEGACY, ConfigurationRoles.INTENDED_RESOLVABLE_BUCKET),
+    @Deprecated
+    LEGACY_TO_INTENDED_CONSUMABLE(ConfigurationRoles.LEGACY, ConfigurationRoles.INTENDED_CONSUMABLE),
+
+    @Deprecated
+    INTENDED_RESOLVABLE_BUCKET_TO_INTENDED_RESOLVABLE(ConfigurationRoles.INTENDED_RESOLVABLE_BUCKET, ConfigurationRoles.INTENDED_RESOLVABLE),
+    @Deprecated
+    INTENDED_CONSUMABLE_BUCKET_TO_INTENDED_CONSUMABLE(ConfigurationRoles.INTENDED_CONSUMABLE_BUCKET, ConfigurationRoles.INTENDED_CONSUMABLE);
+
+    private final boolean consumable;
+    private final boolean resolvable;
+    private final boolean declarableAgainst;
+    private final boolean consumptionDeprecated;
+    private final boolean resolutionDeprecated;
+    private final boolean declarationAgainstDeprecated;
+
+    private final ConfigurationRole initialRole;
+    private final ConfigurationRole eventualRole;
+
+    ConfigurationRolesForMigration(ConfigurationRole initialRole, ConfigurationRole eventualRole) {
+        this.consumable = initialRole.isConsumable();
+        this.resolvable = initialRole.isResolvable();
+        this.declarableAgainst = initialRole.isDeclarableAgainst();
+
+        this.consumptionDeprecated = initialRole.isConsumptionDeprecated() || (initialRole.isConsumable() && !eventualRole.isConsumable());
```

Comment:
Each role encodes deprecations (which it must, to be a role), and these migration roles are roles that map from role to role.  We can't remove the deprecations from the from/to role, since then they wouldn't be roles.

I don't think it's too complicated to say "use the deprecation from the from role, unless it is not deprecated (implied, since we're in the other or branch) but the usage will change from allowed -> disallowed in the eventual role".  I'll add that as an explanatory comment.

It's true we're currently only migrating to `INTENDED` roles which shouldn't have deprecations, at the moment.  I didn't write these roles ahead of time though, I just wrote them one by one as I discovered the need for them by checking each migration.  I'd rather not rely on the migration target being an intended role with no deprecations, and if I checked that, then there would be more logic here.  I'd rather keep it as is, since this calculation is generally correct.